langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java

```diff
@@ -268,40 +276,66 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
     }
 
     @Override
-    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
-        List<Double> queryVector = referenceEmbedding.vectorAsList().stream()
+    public void removeAll() {
+        collection.deleteMany(new Document());
+    }
+
+    @Override
+    public void removeAll(Collection<String> ids) {
+        ensureNotEmpty(ids, "ids");
+        collection.deleteMany(Filters.in("_id", ids));
+    }
+
+    @Override
+    public void removeAll(Filter filter) {
+        ensureNotNull(filter, "filter");
+        collection.deleteMany(map(filter));
+    }
+
+    @Override
+    public EmbeddingSearchResult<TextSegment> search(EmbeddingSearchRequest request) {
+        List<Double> queryVector = request.queryEmbedding().vectorAsList().stream()
                 .map(Float::doubleValue)
                 .collect(toList());
-        long numCandidates = maxResults * maxResultRatio;
-
-        List<Bson> pipeline = Arrays.asList(
-                vectorSearch(
-                        fieldPath("embedding"),
-                        queryVector,
-                        indexName,
-                        numCandidates,
-                        maxResults,
-                        vectorSearchOptions),
-                project(
-                        fields(
-                                metaVectorSearchScore("score"),
-                                include("embedding", "metadata", "text")
-                        )
-                ),
-                match(
-                        Filters.gte("score", minScore)
-                ));
+        long numCandidates = request.maxResults() * maxResultRatio;
+
+        Bson filter = null;
+        if (request.minScore() > 0) {
+            filter = Filters.gte("score", request.minScore());
+        }
+        if (request.filter() != null) {
+            Bson newFilter = map(request.filter());
+            filter = filter == null ? newFilter : Filters.and(filter, newFilter);
+        }
+
+        VectorSearchOptions vectorSearchOptions = this.filter == null
+                ? approximateVectorSearchOptions(numCandidates)
+                : approximateVectorSearchOptions(numCandidates).filter(this.filter);
+
+        ArrayList<Bson> pipeline = new ArrayList<>();
+        pipeline.add(vectorSearch(
+                fieldPath("embedding"),
+                queryVector,
+                indexName,
+                request.maxResults(),
+                vectorSearchOptions));
+        pipeline.add(project(fields(
+                metaVectorSearchScore("score"),
+                include("embedding", "metadata", "text"))));
+        if (filter != null) {
+            Bson match = match(filter);
```

Comment:
[Optional] The request filter currently always applies in post-filtering. If the request doesn’t involve score filtering and all request filter fields are indexed in the index mapping, could we move the filter to the pre-filter stage? We could potentially check this dynamically using `IndexMapping.` However, I'm not sure how common queries without score filtering are.

Reply:
`minScore` should always be present, since it is a double, with null defaulting to 0.

Earlier, I attempted to move most of the filtering into the prefilter (this is the suggested approach), however, I ran into issues and gave this up.